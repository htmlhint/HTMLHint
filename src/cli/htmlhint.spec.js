const ChildProcess = require('child_process')
const run = require('../../test/test-utils').run
const path = require('path')

describe('Executable', () => {
  it('should close stream before exit', (done) => {
    const c = ChildProcess.spawn('node', [
      path.resolve(__dirname, '../bin/htmlhint'),
      '--format',
      'json',
      path.resolve(__dirname, '__fixtures__', 'executable.html'),
    ])
    let stdoutEnd = false
    let processEnd = false
    let isDone = 0

    function checkDone() {
      isDone++
      if (isDone == 2) {
        done()
      }
    }

    c.stdout.on('close', () => {
      stdoutEnd = true
      checkDone()
    })

    c.on('exit', () => {
      processEnd = true
      checkDone()
    })

    c.stdout.on('data', () => {
      expect(stdoutEnd || processEnd).toBe(false)
    })
  })

  for (const format of [
    'checkstyle',
    'compact',
    'default',
    'html',
    'json',
    'junit',
    'markdown',
    'unix',
  ]) {
    it(`should have stdout output with formatter ${format}`, async () => {
      const { exitCode, stdout, stderr } = await run(__dirname, [
        path.resolve(__dirname, '__fixtures__', 'executable.html'),
        '--format',
        format,
      ])

      expect(exitCode).toBe(1)
      expect(stdout).not.toBe('')
      expect(stderr).toBe('')
    })
  }
})
