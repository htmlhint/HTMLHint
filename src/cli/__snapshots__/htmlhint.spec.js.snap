// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Executable should print help with --help: stderr 1`] = `""`;

exports[`Executable should print help with --help: stdout 1`] = `
"Usage: htmlhint <file|folder|pattern|stdin|url ...> [options]

Options:
  -V, --version                                                    output the version number
  -l, --list                                                       show all of the rules available
  -c, --config <file>                                              custom configuration file
  -r, --rules <ruleid, ruleid=value ...>                           set all of the rules available
  -R, --rulesdir <file|folder>                                     load custom rules from file or folder
  -f, --format <checkstyle|compact|html|json|junit|markdown|unix>  output messages as custom format
  -i, --ignore <pattern, pattern ...>                              add pattern to exclude matches
  --nocolor                                                        disable color
  --warn                                                           Warn only, exit with 0
  -h, --help                                                       display help for command
  Examples:

    htmlhint
    htmlhint www
    htmlhint www/test.html
    htmlhint www/**/*.xhtml
    htmlhint www/**/*.{htm,html}
    htmlhint http://www.alibaba.com/
    cat test.html | htmlhint stdin
    htmlhint --list
    htmlhint --rules tag-pair,id-class-value=underline test.html
    htmlhint --config .htmlhintrc test.html
    htmlhint --ignore **/build/**,**/test/**
    htmlhint --rulesdir ./rules/
"
`;

exports[`Executable should print help with -h: stderr 1`] = `""`;

exports[`Executable should print help with -h: stdout 1`] = `
"Usage: htmlhint <file|folder|pattern|stdin|url ...> [options]

Options:
  -V, --version                                                    output the version number
  -l, --list                                                       show all of the rules available
  -c, --config <file>                                              custom configuration file
  -r, --rules <ruleid, ruleid=value ...>                           set all of the rules available
  -R, --rulesdir <file|folder>                                     load custom rules from file or folder
  -f, --format <checkstyle|compact|html|json|junit|markdown|unix>  output messages as custom format
  -i, --ignore <pattern, pattern ...>                              add pattern to exclude matches
  --nocolor                                                        disable color
  --warn                                                           Warn only, exit with 0
  -h, --help                                                       display help for command
  Examples:

    htmlhint
    htmlhint www
    htmlhint www/test.html
    htmlhint www/**/*.xhtml
    htmlhint www/**/*.{htm,html}
    htmlhint http://www.alibaba.com/
    cat test.html | htmlhint stdin
    htmlhint --list
    htmlhint --rules tag-pair,id-class-value=underline test.html
    htmlhint --config .htmlhintrc test.html
    htmlhint --ignore **/build/**,**/test/**
    htmlhint --rulesdir ./rules/
"
`;
